Add more options to copy instance dialog

marcelohdez opened the pull request:
Hello, I've added extra options to the copy instance dialog to not copy certain aspects of the instance like:
* Game options (options.txt)
* Resource packs
* Shaders
* Servers
* Mods ("coremods", "mods", and "config" folders, not mod loader)

And would like to know if this is a worthwhile addition to the project.

This is my first contribution to open source :D, and really my first C++ code outside of school so I apologize in advance if I did something incorrectly -.- let me know and I will fix it. (I've programmed in Java and a little Rust before, also idk how Qt Designer works so I most likely broke something in the .ui file)

*Video showcase:* https://drive.google.com/file/d/1ZSipakoZ0QZAdFY-E2RjdnQVVKDHQouV/view?usp=sharing

marcelohdez committed:
add more options to copy instance dialog

- Copy game options, copy resource packs, copy shaders,
copy servers, and copy mods
- Also made a new InstanceCopyPrefs struct to store those
options rather than passing 7 different booleans into
InstanceCopyTask's constructor

Signed-off-by: Marcelo Hernandez <marcelohdez.inq@gmail.com>

marcelohdez commented:
I could also attempt to add a `Select/Deselect All` button for the checkboxes if wanted

forkiesassds commented:
> I could also attempt to add a `Select/Deselect All` button for the checkboxes if wanted

Would be a great idea

marcelohdez committed:
Add "Select all" checkbox + ui revamp + code cleanup

Signed-off-by: Marcelo Hernandez <marcelohdez.inq@gmail.com>

marcelohdez committed:
Check "Select all" checkbox if all options are already selected

+ code cleanup

Signed-off-by: Marcelo Hernandez <marcelohdez.inq@gmail.com>

marcelohdez commented:
I've added a "Select all" checkbox which gets checked if the other options are already checked and I also gave the UI a two-column view. (Also cleaned up the code)
![GIF showcasing new "Select all" checkbox and two-column view](https://imgur.com/Sr1PEY5.gif)
Also as you can see by the newest commit I also had merge conflicts in CLion 😅... Pain.

flowln commented:
How do you have two identical commits ([`d3d37fc`](https://github.com/PrismLauncher/PrismLauncher/pull/242/commits/d3d37fc8d260c7dd883b709f403f5c299888891c) and [`8692e37`](https://github.com/PrismLauncher/PrismLauncher/pull/242/commits/8692e378faf2c120aa44fb435c9911b6a2d597d5)) :thinking:

marcelohdez commented:
> How do you have two identical commits ([`d3d37fc`](https://github.com/PrismLauncher/PrismLauncher/pull/242/commits/d3d37fc8d260c7dd883b709f403f5c299888891c) and [`8692e37`](https://github.com/PrismLauncher/PrismLauncher/pull/242/commits/8692e378faf2c120aa44fb435c9911b6a2d597d5)) :thinking: 

I attempted to signoff the commit which was already on github then force pushed and ended up creating 2 commits idk how. Should I redo these changes in another branch and open a new pull request? I messed up the commits here cause I didn't know how signing off worked in Git :/

flowln commented:
no need to create another branch, just do `git rebase --signoff -i develop` and drop the commits that shouldn't be here (and it will also sign off you commits :))

marcelohdez commented:
That worked!! Thank you so much! :D I learned something new.

flowln commented on launcher/InstanceCopyPrefs.h:
```diff
@@ -0,0 +1,20 @@
+//
```
Do we need a separate .h/.cpp file for this? maybe we could just have it inside `InstanceCopyTask`

marcelohdez replied:
Okay, I'll work on that!

marcelohdez replied:
Although, if I put it in InstanceCopyTask then we'll have to include InstanceCopyTask in CopyInstanceDialog even though we'll only use that struct and it won't be instantly clear where the struct comes from, does that matter?

flowln replied:
huh, i thought it already included it.
no need to change this then :P

flowln commented on launcher/InstanceCopyTask.cpp:
```diff
@@ -5,18 +5,66 @@
     {
-        // FIXME: get this from the original instance type...
-        auto matcherReal = new RegexpMatcher("[.]?minecraft/saves");
-        matcherReal->caseSensitive(false);
-        m_matcher.reset(matcherReal);
+        appendToFilter(filter, "saves");
     }
+
+    if(!prefs.copyGameOptions) {
+        appendToFilter(filter, "options.txt");
+    }
+
+    if(!prefs.copyResourcePacks)
+    {
+        appendToFilter(filter, "resourcepacks");
+        appendToFilter(filter, "texturepacks");
+    }
+
+    if(!prefs.copyShaderPacks)
+    {
+        appendToFilter(filter, "shaderpacks");
+    }
+
+    if(!prefs.copyServers)
+    {
+        appendToFilter(filter, "servers.dat");
+        appendToFilter(filter, "servers.dat_old");
+        appendToFilter(filter, "server-resource-packs");
+    }
+
+    if(!prefs.copyMods)
+    {
+        appendToFilter(filter, "coremods");
+        appendToFilter(filter, "mods");
+        appendToFilter(filter, "config");
+    }
```
since you made a struct with the preferences, wouldn't it be better to have a function like 'getFiltersFromPreferences' or something in there, that would return a `QStringList` with those filters? Then here, you'd just iterate over the strings on that list, and append them. This way, we don't need to depend on the internals of that struct here

marcelohdez replied:
Okay! I'll work on this

marcelohdez replied:
Okay I added a new `QString getSelectedFiltersAsRegex()` function to InstanceCopyPrefs that looks like:

```C++
// Returns a single RegEx string of the selected folders/files to filter out (ex: ".minecraft/saves|.minecraft/mods")
QString InstanceCopyPrefs::getSelectedFiltersAsRegex() const
{
    QStringList filters;

    if(!copySaves)
        filters << "saves";

    if(!copyGameOptions)
        filters << "options.txt";

    if(!copyResourcePacks)
        filters << "resourcepacks" << "texturepacks";

    if(!copyShaderPacks)
        filters << "shaderpacks";

    if(!copyServers)
        filters << "servers.dat" << "servers.dat_old" << "server-resource-packs";

    if(!copyMods)
        filters << "coremods" << "mods" << "config";

    // If we have any filters to add, join them as a single regex string to return:
    if (!filters.isEmpty()) {
        const QString MC_ROOT = "[.]?minecraft/";
        // Ensure first filter starts with root, then join other filters with OR regex before root (ex: ".minecraft/saves|.minecraft/mods"):
        return MC_ROOT + filters.join("|" + MC_ROOT);
    }

    return "";
}
```

So that the constructor of InstanceCopyTask now looks like this:

```C++
InstanceCopyTask::InstanceCopyTask(InstancePtr origInstance, const InstanceCopyPrefs& prefs)
{
    m_origInstance = origInstance;
    m_keepPlaytime = prefs.keepPlaytime;

    QString filters = prefs.getSelectedFiltersAsRegex();
    if (!filters.isEmpty())
    {
        // Set regex filter:
        // FIXME: get this from the original instance type...
        auto matcherReal = new RegexpMatcher(filters);
        matcherReal->caseSensitive(false);
        m_matcher.reset(matcherReal);
    }
}
```

Is this good?

flowln replied:
sure, thx!

flowln commented on launcher/ui/dialogs/CopyInstanceDialog.cpp:
```diff
@@ -78,8 +78,13 @@ CopyInstanceDialog::CopyInstanceDialog(InstancePtr original, QWidget *parent)
     ui->groupBox->lineEdit()->setPlaceholderText(tr("No group"));
-    ui->copySavesCheckbox->setChecked(m_copySaves);
-    ui->keepPlaytimeCheckbox->setChecked(m_keepPlaytime);
+    ui->copySavesCheckbox->setChecked(m_selectedOptions.copySaves);
+    ui->keepPlaytimeCheckbox->setChecked(m_selectedOptions.keepPlaytime);
+    ui->copyGameOptionsCheckbox->setChecked(m_selectedOptions.copyGameOptions);
+    ui->copyResPacksCheckbox->setChecked(m_selectedOptions.copyResourcePacks);
+    ui->copyShaderPacksCheckbox->setChecked(m_selectedOptions.copyShaderPacks);
+    ui->copyServersCheckbox->setChecked(m_selectedOptions.copyServers);
+    ui->copyModsCheckbox->setChecked(m_selectedOptions.copyMods);
```
maybe we could have getters / setters here, to further abstract away the internal data from that struct. may be a bit overkill though, so feel free to resolve this if you wish :p

marcelohdez replied:
Okay, I'll start work on this soon

marcelohdez replied:
Done! commit c00f96c.

flowln commented on launcher/ui/dialogs/CopyInstanceDialog.cpp:
```diff
@@ -117,6 +122,40 @@ QString CopyInstanceDialog::instGroup() const
+// Sets b to true if state is a checked checkbox
+void CopyInstanceDialog::checkBool(bool& b, const int& state)
+{
+    if(state == Qt::Unchecked)
+    {
+        b = false;
+    }
+    else if(state == Qt::Checked)
+    {
+        b = true;
+    }
```
```suggestion
    b = (state == Qt::Checked);
```

my change assumes `state` is not a tristate (at least from reading the function comment). Is that true?

marcelohdez replied:
Yes, the checkbox is not tristate, although the previous implementation had used those two if statements as well so I kept them in case it meant anything, I could change it to that single line

flowln commented on launcher/ui/dialogs/CopyInstanceDialog.cpp:
```diff
@@ -117,6 +122,40 @@ QString CopyInstanceDialog::instGroup() const
+}
+
+// Sets b to true if state is a checked checkbox
+void CopyInstanceDialog::checkBool(bool& b, const int& state)
```
```suggestion
bool CopyInstanceDialog::checkBool(int state)
```

also change the header!

marcelohdez replied:
I wrote the header this way because I'm updating the bool inside, then checking if everything in m_selectedOptions is true, and finally checking the "Select all" check box accordingly. I thought it'd be better to do this all within the same function since we need to update the Select all checkbox every time we change a value in m_selectOptions anyways, so we'd save some code.

Currently:
```C++
checkBool(m_selectedOptions.copySaves, state); // Update copySaves and select all checkbox
```

If I change it to return a new boolean instead, I would need to update the select all checkbox separately once the value has changed in m_selectOptions (unless there's a simpler way to do this) so It would be something like:

```C++
m_selectedOptions.copySaves = checkBool(state); // Update copySaves
updateSelectAllCheckbox(); // <-- new function
```

Is the new one prefereable? (So I know in the future)

flowln replied:
i think it's best to separate those two. messing with argument references tends to be harder to read and more error-prone then simply returning a value, and then assigning it to whatever variable you want.

I'd also change `checkBool` to a more descriptive name now :p (actually, do you really need a separate function just for that? :thinking: )

marcelohdez replied:
Okay! And yeah a function is not really needed, also should the updateSelectAllCheckbox() function be in the private slots area of the CopyInstanceDialog class? Or should it be in the regular private section with the data?

flowln replied:
doesn't really matter. since you're not connecting any signals to it, a non-slot works just fine

marcelohdez replied:
Done! commit 385c452

flowln reviewed (commented):
main thing here is that perhaps it would be a good idea to abstract some stuff away to ease maintenance in the future, not that much though.

the feature is cool, thanks for working on it :D

marcelohdez committed:
Simplify bool check in CopyInstanceDialog.cpp

Signed-off-by: Marcelo Hernandez <marcelohdez.inq@gmail.com>

marcelohdez committed:
remove checkBool function, add updateSelectAllCheckbox function

Signed-off-by: Marcelo Hernandez <marcelohdez.inq@gmail.com>

marcelohdez committed:
Abstract away InstanceCopyPrefs' internals through new getSelectedFiltersAsRegex() function

+ fix typo in comment
+ remove unused import
+ add [[nodiscard]] to methods

Signed-off-by: Marcelo Hernandez <marcelohdez.inq@gmail.com>

marcelohdez committed:
create getters and setters for InstanceCopyPrefs

+ use pragma once like other .h files in this directory

Signed-off-by: Marcelo Hernandez <marcelohdez.inq@gmail.com>

marcelohdez commented:
I was wondering, since we are already here, should I also update the check boxes from using `on_foo_bar` slots to the `QObject::connect` API? or should that be a separate PR? Additionally, I noticed I could also add an option to not copy screenshots, although 9 checkboxes may be too much visual clutter, let me know.

flowln reviewed (approved):
This looks good to me, thank you!

> I was wondering, since we are already here, should I also update the check boxes from using on_foo_bar slots to the QObject::connect API? or should that be a separate PR?

I don't think it warrants a separate PR, both work just fine, it's just that the `on_foo_bar` method is sometimes considered bad practice and should be avoided when possible. When working on older code, though, no need to change it, and it's better to keep consistency. Anyway, change it to the other way if you want to, but don't feel obligated to.

> Additionally, I noticed I could also add an option to not copy screenshots, although 9 checkboxes may be too much visual clutter, let me know.

I too think it's getting very cluttered in there. I don't mind either way, but if you do add, you can put the 'Select all' checkbox on top, horizontally centered, and then have a 2x4 grid of checkboxes for the options, since adding one more would make for an odd number of options.

marcelohdez committed:
added option to not copy screenshots

+ moved select all checkbox to top row, centered.

Signed-off-by: Marcelo Hernandez <marcelohdez.inq@gmail.com>

marcelohdez commented:
Okay! I added the copy screenshots checkbox, and centered the select all checkbox on the top row. For the sake of consistency I'll keep the `on_foo_bar` slots, so this pull request *should* be done :D

Scrumplex commented on launcher/InstanceCopyPrefs.h:
```diff
@@ -0,0 +1,41 @@
+   public:
+    [[nodiscard]] bool allTrue() const;
+    [[nodiscard]] QString getSelectedFiltersAsRegex() const;
+    // Getters
+    [[nodiscard]] bool isCopySavesEnabled() const;
+    [[nodiscard]] bool isKeepPlaytimeEnabled() const;
+    [[nodiscard]] bool isCopyGameOptionsEnabled() const;
+    [[nodiscard]] bool isCopyResourcePacksEnabled() const;
+    [[nodiscard]] bool isCopyShaderPacksEnabled() const;
+    [[nodiscard]] bool isCopyServersEnabled() const;
+    [[nodiscard]] bool isCopyModsEnabled() const;
+    [[nodiscard]] bool isCopyScreenshotsEnabled() const;
+    // Setters
+    void enableCopySaves(bool b);
+    void enableKeepPlaytime(bool b);
+    void enableCopyGameOptions(bool b);
+    void enableCopyResourcePacks(bool b);
+    void enableCopyShaderPacks(bool b);
+    void enableCopyServers(bool b);
+    void enableCopyMods(bool b);
+    void enableCopyScreenshots(bool b);
```
I don't think we need getters and setters here. As it's just a struct, we can make the fields public. Should reduce the code a bit.

flowln replied:
heh i asked them to make getters and setters :p

having them should make it easier to change the struct internals if we ever need to

Scrumplex replied:
> having them should make it easier to change the struct internals if we ever need to

Fair, but I don't know if we would ever change the internal much though :eyes:

flowln replied:
better safe than sorry :p ~i need to somehow justify the work i made them do~

Scrumplex replied:
I guess it's fine then. Can't promise that I would never remove this during a refactor though :eyes:

Scrumplex reviewed (approved):
Seems to work fine otherwise :+1:

Scrumplex commented:
Thanks!
